sklearn/datasets/openml.py

```diff
@@ -45,13 +45,32 @@ def _open_openml_url(openml_path, data_home):
         Directory to which the files will be cached. If None, no caching will
         be applied.
 
+    md5_checksum: str
+        Checksum of the OpenML ressource downloaded. If None, no verification
+        is done.
+
     Returns
     -------
     result : stream
         A stream to the OpenML resource
     """
     if data_home is None:
-        return urlopen(_OPENML_PREFIX + openml_path)
+
+        response = urlopen(_OPENML_PREFIX + openml_path)
+        content = response.read()
+
+        if md5_checksum is not None:
+            md5_hash = hashlib.md5(content).hexdigest()
+            if md5_checksum != md5_hash:
+                warn('Data set file hash {} does not match the checksum {}.'
+                     .format(md5_hash, md5_checksum))
+
+        fp = tempfile.TemporaryFile()
```

Comment:
It would be preferable to avoid creating a temporary file. Maybe using `BytesIO` ?